File: src/3sum-closest.py
"""
首先如果数组长度小于等于3，则直接返回数组的和，用了reduce，好棒的感觉！

排序数组，使其从大到小排列，方便后面缩小范围

双重循环，第一重遍历数组，作为第一个数
第二重，设置两个指针，分别指向剩余数组的头和尾，
将 i j k 位置的和与result进行比较
最后如果 current_sum 大于 target，则需要缩小大数范围，k--
反之则j++
"""

class Solution:
    # @return an integer
    def threeSumClosest(self, num, target):
        length = len(num)
        if length <= 3:
            return reduce(lambda x,y: x+y, num)

        num.sort()
        result = num[0] + num[1] + num[2]

        for i in xrange(length-2):
            j = i + 1
            k = length - 1
            while j < k:
                current_sum = num[i] + num[j] + num[k]
                if abs(current_sum - target) < abs(result - target):
                    result = current_sum
                    if result == target:
                        return result

                if current_sum > target:
                    k -= 1
                else:
                    j += 1

        return result


# test
print Solution().threeSumClosest([-1,2,1,-4],1)
print Solution().threeSumClosest([1,2,3],1)
print Solution().threeSumClosest([1,2],1)
print Solution().threeSumClosest([1],1)
print Solution().threeSumClosest([],1)

File: cpp/3sum-closest.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int length = nums.size();
        if (length < 3) {
            return 0;
        }

        sort(nums.begin(), nums.end());
        int closet = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < length - 2; i++) {
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }

            int p = i + 1, q = length - 1;
            while (p < q) {
                int curSum = nums[i] + nums[p] + nums[q];
                if (curSum == target) {
                    return target;
                } else {
                    if (abs(target - curSum) < abs(target - closet)) {
                        closet = curSum;
                    }
                    if (curSum > target) {
                        --q; 
                    } else {
                        ++p;
                    }
                }
            }
        }

        return closet;
    }
};

int main () {
    vector<int> nums {-1, 2, 1, -4};
    int target = 1;

    Solution s;
    cout << s.threeSumClosest(nums, target) << endl;


    return 0;
}
